File: apps/api/src/ask/byollm.ts
// Per-request BYOLLM lane for `/v1/ask` — the apps/api half of
// `SK-LLM-016` step 1 (the `x-nlq-byollm-key` header). The pure
// precedence + provider primitives live in `@nlqdb/llm`
// (`SK-LLM-019`/`SK-LLM-020`); this module parses + authorizes the header
// into a `ByollmCredential` and resolves which `LLMRouter` the ask
// pipeline runs on, so the wire format and the signed-in-only gate live
// in one tested place rather than inline in the route handler.
//
// Scope: header lane only. Account-stored keys (`api_keys.scope="byollm"`,
// KEK-decrypt) and the hosted-premium chain (`SK-LLM-017`, dark pre-§6)
// are not wired here — both stay on the free router. Surface parity
// (SDK/CLI/MCP/elements) is tracked as a gap in `premium-tier/FEATURE.md`
// per `GLOBAL-003`.

import {
  type ByollmCredential,
  buildByollmRouter,
  dispatchLaneAttributes,
  type LLMRouter,
  selectDispatchLane,
} from "@nlqdb/llm";

export type { ByollmCredential } from "@nlqdb/llm";

// Canonical header name (`SK-PREMIUM-008`). Lower-case — Hono normalises
// header lookups, but keeping the constant lower-case matches the wire.
export const BYOLLM_HEADER = "x-nlq-byollm-key";

// AI Gateway compat-endpoint provider slugs we accept for BYOLLM. Pinned
// to the `<provider>/<model>` slugs the unified `/compat/chat/completions`
// endpoint actually serves (verified against developers.cloudflare.com/
// ai-gateway, 2026-05): `openai`, `anthropic`, `google-ai-studio`.
// `SK-PREMIUM-008` also lists OpenRouter, but it is not a compat-endpoint
// provider — tracked as a gap in `premium-tier/FEATURE.md`. An unknown
// slug fails loud here (GLOBAL-012) rather than 404-ing at the gateway.
export const SUPPORTED_BYOLLM_PROVIDERS = ["openai", "anthropic", "google-ai-studio"] as const;
const SUPPORTED = new Set<string>(SUPPORTED_BYOLLM_PROVIDERS);

// One-sentence header-shape message (GLOBAL-012), reused by the malformed
// branches so the contract reads identically wherever it surfaces.
const SHAPE_MESSAGE = `${BYOLLM_HEADER} must be "<provider>:<model>:<key>" (e.g. "openai:gpt-5.2:sk-…").`;

export type ParseByollmResult =
  // `credential: null` means the header was absent or blank — no BYOLLM
  // intent, dispatch falls through to the free chain.
  { ok: true; credential: ByollmCredential | null } | { ok: false; message: string };

// Parse the `x-nlq-byollm-key` header into a credential. Pure + I/O-free.
// Format: `<provider>:<model>:<key>`. We split on the first two colons
// only and take the remainder as the key, so a key containing a colon
// survives intact; provider slugs and the supported providers' model ids
// never contain a colon. Fails loud on a malformed value or an
// unsupported provider (GLOBAL-012) — a confusing upstream 4xx is worse
// than an obvious 400 at the edge.
export function parseByollmHeader(raw: string | undefined): ParseByollmResult {
  if (raw === undefined) return { ok: true, credential: null };
  const trimmed = raw.trim();
  if (trimmed === "") return { ok: true, credential: null };

  const firstColon = trimmed.indexOf(":");
  const secondColon = firstColon === -1 ? -1 : trimmed.indexOf(":", firstColon + 1);
  if (firstColon === -1 || secondColon === -1) {
    return { ok: false, message: SHAPE_MESSAGE };
  }

  const upstream = trimmed.slice(0, firstColon).trim().toLowerCase();
  const model = trimmed.slice(firstColon + 1, secondColon).trim();
  const apiKey = trimmed.slice(secondColon + 1).trim();
  if (model === "" || apiKey === "") return { ok: false, message: SHAPE_MESSAGE };
  if (!SUPPORTED.has(upstream)) {
    return {
      ok: false,
      message: `BYOLLM provider "${upstream}" is not supported; use one of ${SUPPORTED_BYOLLM_PROVIDERS.join(", ")}.`,
    };
  }
  return { ok: true, credential: { apiKey, upstream, model } };
}

export type ResolveAskRouterResult =
  | { ok: true; router: LLMRouter; attributes: Record<string, string> }
  // The deployment has a BYOLLM key inbound but no AI Gateway configured
  // — an operator-config gap, surfaced as 503 by the caller (not 4xx: the
  // request is well-formed, the platform just can't serve the lane).
  | { ok: false; reason: "gateway_unconfigured" };

// Resolve the ask-pipeline router from the per-request credential.
// Header credential → BYOLLM router through the user's own key; otherwise
// the free router. (Account-stored + premium lanes aren't wired this
// slice; `selectDispatchLane` still owns the precedence so adding them is
// a one-field change.) Returns the redacted `llm.dispatch_lane` span
// attributes alongside, so the caller annotates the existing ask span
// without a second source of truth. Pure + I/O-free.
export function resolveAskRouter(args: {
  headerCredential: ByollmCredential | null;
  freeRouter: LLMRouter;
  gateway: { accountId?: string; gatewayId?: string };
  userId: string;
}): ResolveAskRouterResult {
  const selection = selectDispatchLane({ headerCredential: args.headerCredential });
  const attributes = dispatchLaneAttributes(selection);
  if (selection.lane !== "byollm") {
    return { ok: true, router: args.freeRouter, attributes };
  }
  const { accountId, gatewayId } = args.gateway;
  if (!accountId || !gatewayId) return { ok: false, reason: "gateway_unconfigured" };
  const router = buildByollmRouter({
    credential: selection.credential,
    accountId,
    gatewayId,
    userId: args.userId,
  });
  return { ok: true, router, attributes };
}

File: cli/internal/byollm/byollm.go
// Package byollm parses and validates a bring-your-own-LLM provider key
// into the `x-nlq-byollm-key` wire value, mirroring the TS SDK's single
// tested helper (SK-SDK-010) so the colon-join hazard lives in one place
// across surfaces (GLOBAL-002). The key is dispatched at 0% markup per
// GLOBAL-026; the lane is signed-in only (SK-LLM-021), enforced by the
// server — this package only owns the wire shape.
package byollm

import (
	"fmt"
	"strings"
)

// HeaderName is the wire header the API consumes (SK-LLM-021). Lower-case
// to match the canonical name in apps/api and the TS SDK.
const HeaderName = "x-nlq-byollm-key"

// Credential is a validated BYOLLM provider key. Provider is lower-cased;
// Model is the raw upstream id (BYOLLM is the escape hatch where the user
// owns the model choice, unlike the hosted `model` preset).
type Credential struct {
	Provider string
	Model    string
	Key      string
}

// Header returns the `<provider>:<model>:<key>` value. Safe because Parse
// guarantees Provider and Model are colon-free; only Key may contain a
// colon, and it is the unsplit remainder the server takes verbatim.
func (c Credential) Header() string {
	return c.Provider + ":" + c.Model + ":" + c.Key
}

// Redacted renders the credential for `nlq byollm status` without leaking
// the key — provider/model in clear, key as `…<last 4>`.
func (c Credential) Redacted() string {
	return fmt.Sprintf("%s · %s · key %s", c.Provider, c.Model, maskKey(c.Key))
}

// Parse validates and normalises the three parts, mirroring the SDK's
// buildByollmHeader: provider lower-cased, all parts non-empty,
// provider/model colon-free (the key may contain `:` as the unsplit
// remainder), and no part may hold a control char (CR/LF would smuggle a
// second header). Fails loud with a one-sentence message (GLOBAL-012).
func Parse(provider, model, key string) (Credential, error) {
	p := strings.ToLower(strings.TrimSpace(provider))
	m := strings.TrimSpace(model)
	k := strings.TrimSpace(key)
	if p == "" || m == "" || k == "" {
		return Credential{}, fmt.Errorf("provider, model, and key must all be non-empty")
	}
	if strings.ContainsRune(p, ':') || strings.ContainsRune(m, ':') {
		return Credential{}, fmt.Errorf("provider and model must not contain a colon")
	}
	for _, r := range p + m + k {
		if r < 0x20 || r == 0x7f {
			return Credential{}, fmt.Errorf("provider, model, and key must not contain control characters")
		}
	}
	return Credential{Provider: p, Model: m, Key: k}, nil
}

// FromStored reconstructs a Credential from the joined header value kept
// in the credential store. Splits on the first two colons only so a key
// containing a colon survives intact. Returns false when the stored value
// is missing or malformed (so callers treat a corrupt slot as "unset"
// rather than dispatching a guaranteed-400 header).
func FromStored(value string) (Credential, bool) {
	first := strings.IndexByte(value, ':')
	if first < 0 {
		return Credential{}, false
	}
	second := strings.IndexByte(value[first+1:], ':')
	if second < 0 {
		return Credential{}, false
	}
	second += first + 1
	cred, err := Parse(value[:first], value[first+1:second], value[second+1:])
	if err != nil {
		return Credential{}, false
	}
	return cred, true
}

func maskKey(key string) string {
	if len(key) <= 4 {
		return "****"
	}
	return "…" + key[len(key)-4:]
}
